armi/reactor/blocks.py

```diff
@@ -2009,10 +2049,36 @@ def rotate(self, rad):
 
         """
         rotNum = round((rad % (2 * math.pi)) / math.radians(60))
-        self._rotatePins(rotNum)
+        self._rotateChildLocations(rad, rotNum)
+        self.p.orientation[2] += rotNum * 60
         self._rotateBoundaryParameters(rotNum)
         self._rotateDisplacement(rad)
 
+    def _rotateChildLocations(self, radians: float, rotNum: int):
+        if self.spatialGrid is None:
+            return
+        locationRotator = functools.partial(
+            self.spatialGrid.rotateIndex, rotations=rotNum
+        )
+        rotationMatrix = np.array(
+            [
+                [math.cos(radians), -math.sin(radians)],
+                [math.sin(radians), math.cos(radians)],
+            ]
+        )
+        for c in self:
+            if isinstance(c.spatialLocator, grids.MultiIndexLocation):
+                newLocations = list(map(locationRotator, c.spatialLocator))
+                c.spatialLocator = grids.MultiIndexLocation(self.spatialGrid)
+                c.spatialLocator.extend(newLocations)
+            elif isinstance(c.spatialLocator, grids.CoordinateLocation):
+                oldCoords = c.spatialLocator.getLocalCoordinates()
+                newXY = rotationMatrix.dot(oldCoords[:2])
+                newLocation = grids.CoordinateLocation(
+                    newXY[0], newXY[1], oldCoords[2], self.spatialGrid
+                )
+                c.spatialLocator = newLocation
```

Comment:
There's no "else" clause here. Hmmm... Usually, it's good practice to catch those. Raise a warning or an error.

I guess in this location the question is what if the component has multiplicity 1, so it just has the default `IndexLocation`?  We just don't  want to rotate that because we assume it is at the center of the block? Is that your idea?

Reply:
Yeah I don't love this either. It looks if a component is placed on a grid, even if it has a single location, it **always** gets a `MultiIndexLocator`

https://github.com/terrapower/armi/blob/a3ef1a39325e2f33de966f8d11a7e7c4d380f3a3/armi/reactor/blueprints/blockBlueprint.py#L163-L182

https://github.com/terrapower/armi/blob/a3ef1a39325e2f33de966f8d11a7e7c4d380f3a3/armi/reactor/blueprints/gridBlueprint.py#L473-L477

The `CoordinateLocation` gets set during `HexBlock.autoCreateSpatialGrids`

https://github.com/terrapower/armi/blob/a3ef1a39325e2f33de966f8d11a7e7c4d380f3a3/armi/reactor/blocks.py#L2372-L2376

So we should have a `MultiIndexLocation` with one or more entries, or a `CoordinateLocation`. I'll make sure this raises the appropriate error.

Reply:
Some [testing ](https://github.com/terrapower/armi/actions/runs/11333020227/job/31516341360#step:6:1126)uncovered an `IndexLocation`
```
TypeError: <DerivedShape: coolant> on <fuel B-2307749255035-001 at ExCore XS: A BU GP: A> has an invalid spatial locator for rotation: <IndexLocation @ (0,0,0)>
```
Added logic to handle that case as well